### PR TITLE
#179 Consistently use URI object in APIs

### PR DESCRIPTION
**Pull request by ndoschek**
Consistently use URI object in APIs

- Use `urijs` as typed URI library
- Use URI objects in APIs and internally for proper URI handling
  - Adapt interfaces, tests
- Customize JsonRpcProxyFactory to ensure URI object is correctly typed onRequest/onNotifciation
  - Configure on backend and frontend for the JsonRpc WebSocketConnection
- Rename modeluri to lowercase where applicable for better alignment
- Differentiate URI and TheiaURI object
- Update README examples

Update dev-example to ModelServerClient V2

- Update custom DevModelServerClient
- Update api-test-menu
  - User URIs to trigger commands
  - Refine notification printing
  - Add test commands that edit via JSON patch commands
  - Remove xmi test commands as they are not relevant anymore
 
General
- Remove simple-import-sort eslint plugin as it interferes with the built-in import sort
- Ignore zipped rolling log files
- Fix launch configs

Part of https://github.com/eclipse-emfcloud/emfcloud/issues/179

Contributed on behalf of STMicroelectronics

### CODE REVIEW

File: examples/dev-example/src/node/model-server-client.ts [hunk 1]
```diff
@@ -8,46 +8,44 @@
...
 
 import { DevModelServerClient } from '../common/dev-model-server-client';
 
 @injectable()
-export class CustomDevModelServerClient extends TheiaBackendModelServerClient implements DevModelServerClient {
+export class CustomDevModelServerClient extends TheiaBackendModelServerClientV2 implements DevModelServerClient {
     private intervalId: NodeJS.Timeout;
 
-    subscribe(modeluri: string, options: SubscriptionOptions = {}): void {
+    subscribe(modeluri: URI, listener?: SubscriptionListener, options: SubscriptionOptionsV2 = {}): SubscriptionListener {
```
Comment (eneufeld) [inline examples/dev-example/src/node/model-server-client.ts:28]: We had multiple examples where the options were set but the listener was undefined. Why use subscription options if no listener is set anyway? This happened in `examples/dev-example/src/browser/api-test-menu.ts`

File: examples/dev-example/src/node/model-server-client.ts [hunk 2]
```diff
@@ -8,46 +8,44 @@
...
     }
 
-    private setKeepAliveInterval(modelUri: string, timeout: number): void {
-        if (!this.isSocketOpen(modelUri) && modelUri === 'Coffee.ecore') {
+    private setKeepAliveInterval(modeluri: URI, timeout: number): void {
+        if (!this.isSocketOpen(modeluri) && modeluri.toString() === 'Coffee.ecore') {
             this.intervalId = setInterval(
-                () => this.send(modelUri, { type: MessageType.keepAlive, data: undefined }),
-                timeout > 1000 ? timeout - 1000 : 1
+                () => this.send(modeluri, { type: MessageType.keepAlive, data: '' }),
```
Comment (eneufeld) [inline examples/dev-example/src/node/model-server-client.ts:38]: why do we send an empty string as data instead of `undefined` now?

File: examples/dev-example/src/browser/api-test-menu.ts [hunk 3]
```diff
@@ -598,180 +602,169 @@ export class ApiTestMenuContribution implements MenuContribution, CommandContrib
...
         commands.registerCommand(SubscribeCommand, {
             execute: () => {
-                this.modelServerClient.subscribe('SuperBrewer3000.coffee');
+                this.modelServerClient.subscribe(new URI('SuperBrewer3000.coffee'), undefined, { errorWhenUnsuccessful: true });
             }
         });
         commands.registerCommand(SubscribeWithValidationCommand, {
             execute: () => {
-                this.modelServerClient.subscribe('SuperBrewer3000.coffee', { livevalidation: true });
+                this.modelServerClient.subscribe(new URI('SuperBrewer3000.coffee'), undefined, { livevalidation: true });
```
Comment (eneufeld) [inline examples/dev-example/src/browser/api-test-menu.ts:723]: we have no listener, why do we need subscription options?

File: packages/modelserver-client/src/model-server-client-v2-integration.spec.ts [hunk 4]
```diff
@@ -234,7 +240,7 @@ describe('Integration tests for ModelServerClientV2', () => {
             const notification = await patchNotification;
             const patch = notification.patch;
 
-            expect(notification.modelUri).to.be.equal(modeluri);
+            expect(notification.modeluri.toString()).to.be.equal(modeluri.toString());
```
Comment (eneufeld) [inline packages/modelserver-client/src/model-server-client-v2-integration.spec.ts:243]: maybe we should use this for comparison: http://medialize.github.io/URI.js/docs.html#equals

File: packages/modelserver-client/src/model-server-client-v2.spec.ts [hunk 5]
```diff
@@ -36,19 +42,19 @@ describe('tests for ModelServerClientV2', () => {
 
     it('initialize - correct baseUrl config of axios instance', () => {
         const axios = client['restClient'];
-        expect(axios.defaults.baseURL).to.be.equal(baseUrl);
+        expect(axios.defaults.baseURL).to.be.equal(baseUrl.toString());
```
Comment (eneufeld) [inline packages/modelserver-client/src/model-server-client-v2.spec.ts:45]: as above maybe use the internal equals method?

File: packages/modelserver-client/src/model-server-client-v2.spec.ts [hunk 6]
```diff
@@ -362,23 +368,23 @@ describe('tests for ModelServerClientV2', () => {
...
             const model1: Model = {
-                modelUri: 'path/to/model1',
+                modeluri: 'path/to/model1',
                 content: {
                     name: 'coffee'
                 }
             };
             const model2: Model = {
-                modelUri: 'path/to/model2',
+                modeluri: 'path/to/model2',
```
Comment (eneufeld) [inline packages/modelserver-client/src/model-server-client-v2.spec.ts:379]: why did you change `modelUri` to `modeluri`?

**Review by eneufeld — COMMENTED** [review]
Code looks good overall. I have some questions and remarks, though.

File: examples/dev-example/src/node/model-server-client.ts (hunk 1 above)
Comment (ndoschek) [reply examples/dev-example/src/node/model-server-client.ts:28]: Previously with `v1`, the listener only was a member of the `SubscriptionOptions`. With `v2` it should be a separate argument (for compatibility reasons, the options member will be used as the listener internally then).

> Why use subscription options if no listener is set anyway?

Yes I agree that it might look a bit off, but in the case of the `api-test-menu` example we make use of a `ModelServerSubscriptionClient` that is initalized once and shows notifications received by opened subscriptions.

Perhaps we could re-work the example application a bit (in a separate ticket) and align the listener usage and also show the request (data like command or patch).

File: examples/dev-example/src/node/model-server-client.ts (hunk 2 above)
Comment (ndoschek) [reply examples/dev-example/src/node/model-server-client.ts:38]: Good catch thanks, this is an leftover can be reverted again.

File: packages/modelserver-client/src/model-server-client-v2-integration.spec.ts (hunk 4 above)
Comment (ndoschek) [reply packages/modelserver-client/src/model-server-client-v2-integration.spec.ts:243]: Great idea, thanks for catching that, will update in upcoming commit.

File: packages/modelserver-client/src/model-server-client-v2.spec.ts (hunk 5 above)
Comment (ndoschek) [reply packages/modelserver-client/src/model-server-client-v2.spec.ts:45]: Thanks, will update those comparisons as well in upcoming commit.

File: packages/modelserver-client/src/model-server-client-v2.spec.ts (hunk 6 above)
Comment (ndoschek) [reply packages/modelserver-client/src/model-server-client-v2.spec.ts:379]: We basically used `modeluri` wherever possible, there were some leftovers especially in `v2` with used `modelUri`, I aligned everything to lowercase (to match the style of the actual used parameter).
There is one place left, where we use camel case, this is due to the data structure the model server sends, I guess this could be aligned at some point on the Java side to avoid any conflicts.

File: packages/modelserver-client/src/model-server-client-api-v1.ts [hunk 7]
```diff
@@ -28,7 +30,7 @@ export interface ModelServerClientApiV1 {
      * @param defaultFormat Optional fallback format that should used when a request method is invoked and no explicit format argument
      *                      has been passed.
      */
-    initialize(baseUrl: string, defaultFormat?: string): void | Promise<void>;
+    initialize(baseUrl: URI, defaultFormat?: string): void | Promise<void>;
```
Comment (cdamus) [inline packages/modelserver-client/src/model-server-client-api-v1.ts:33]: I'm not sure about using the `URI` type for the HTTP resource/request URL. It's not really exactly an URI. Perhaps instead this should be the Node built-in [`URL`](https://nodejs.org/api/url.html) type. What do you think?

File: packages/modelserver-client/src/model-server-client-api-v2.ts [hunk 8]
```diff
@@ -70,14 +71,14 @@ export interface ModelServerClientApiV2 {
...
-     * (e.g. `http://localhost:8081/api/v1/`). Once the initialization is completed the client is expected
+     * (e.g. `http://localhost:8081/api/v2/`). Once the initialization is completed the client is expected
      * to be ready and should be able to handle REST requests to & responses from the model server.
      * Any requests to the model server before the client has been initialized are expected to fail (i.e. throw an error)
      * @param baseUrl Url pointing to the API entry point
      * @param defaultFormat Optional fallback format that should used when a request method is invoked and no explicit format argument
      *                      has been passed. The default-default is `'json-v2'`
      */
-    initialize(baseUrl: string, defaultFormat?: Format): void | Promise<void>;
+    initialize(baseUrl: URI, defaultFormat?: Format): void | Promise<void>;
```
Comment (cdamus) [inline packages/modelserver-client/src/model-server-client-api-v2.ts:81]: Same comment as above about `URL` type. This obviously would cascade to call sites elsewhere in this PR.

File: packages/modelserver-client/src/model-server-client-v2-integration.spec.ts [hunk 9]
```diff
@@ -234,7 +240,7 @@ describe('Integration tests for ModelServerClientV2', () => {
             const notification = await patchNotification;
             const patch = notification.patch;
 
-            expect(notification.modelUri).to.be.equal(modeluri);
+            expect(notification.modeluri.equals(modeluri)).to.be.true;
```
Comment (cdamus) [inline packages/modelserver-client/src/model-server-client-v2-integration.spec.ts:243]: My only quibble with this is that it loses the presentation of the string comparison when the test fails, which is highly informative in the test report.

Is there a good reason not to test equality of the toString() conversions of the URIs? I wouldn't go so far as to implement a Chai assertion just for equality testing of URIs ... 😀 

(there are many more occurrences of this throughout the tests)

File: packages/modelserver-client/src/model-server-client-v2-integration.spec.ts [hunk 10]
```diff
@@ -399,14 +405,10 @@ describe('Integration tests for ModelServerClientV2', () => {
...
             patchedMachine.children[1].processor.clockSpeed = 6;
 
-            // Generate patch by diffing the original model and the patched one
-            const patch = jsonpatch.compare(machine, patchedMachine);
-            const modelPatch = {
-                modelUri: modeluri,
-                patch
-            };
+            // Generate patches by diffing the original model and the patched one
+            const patches = jsonpatch.compare(machine, patchedMachine);
```
Comment (cdamus) [inline packages/modelserver-client/src/model-server-client-v2-integration.spec.ts:409]: Nit: an array of `Operation` is a patch.
```suggestion
            const patch = jsonpatch.compare(machine, patchedMachine);
```

**Review by cdamus — COMMENTED** [review]
Looks great! Thanks. I have just a few nits to remark on, but possibly not significant enough to need changes.

File: examples/dev-example/src/node/model-server-client.ts (hunk 1 above)
Comment (eneufeld) [reply examples/dev-example/src/node/model-server-client.ts:28]: Could you please open a follow up here as suggested by you?

File: packages/modelserver-client/src/model-server-client-v2-integration.spec.ts (hunk 9 above)
Comment (eneufeld) [reply packages/modelserver-client/src/model-server-client-v2-integration.spec.ts:243]: Looking at the [docs](http://medialize.github.io/URI.js/docs.html#equals), the equals method does things like normalization and also ignores the order of query parameters. This is probably not needed here in the tests, but still cleaner imho.
The `expect` accepts a second parameter, which we should use to provide an error message which is helpful.
@cdamus, @ndoschek what do you think?

File: packages/modelserver-client/src/model-server-client-api-v1.ts (hunk 7 above)
Comment (eneufeld) [reply packages/modelserver-client/src/model-server-client-api-v1.ts:33]: I agree, the baseUrl is actually an URL (identifying the endpoint **location** as our jsdoc states) and not an URI, so using the URL type makes sense here.

File: examples/coffee-theia/src/node/backend-module.ts [hunk 11]
```diff
@@ -18,7 +18,7 @@ export default new ContainerModule(bind => {
 
 @injectable()
 export class SimpleLaunchOptions implements LaunchOptions {
-    baseURL = 'api/v1/';
+    baseURL = 'api/v2';
```
[the remark this answers is not in the document]
Comment (ndoschek) [reply examples/coffee-theia/src/node/backend-module.ts:21]: In general, URI building is very stable now for both versions, as we use library functions to add paths and and fragments (e.g. [`segment()`](https://medialize.github.io/URI.js/docs.html#accessors-segment)) which handle the "slash" handling automatically. So in the end it does not really matter if the base has a trailing slash or not.
I just aligned this baseUrl here with the others (and we also did that for the modelserver-node already as well).

We also have a test that covers that just in case: [`'test createSubscriptionPath with/without trailing slash'`](https://github.com/eclipse-emfcloud/emfcloud-modelserver-theia/blob/ac65d3dc16817997d738aa1cbb3a82de93822000/packages/modelserver-client/src/model-server-client-v2.spec.ts#L41)

File: packages/modelserver-client/src/model-server-client-v2-integration.spec.ts (hunk 4 above)
Comment (ndoschek) [reply packages/modelserver-client/src/model-server-client-v2-integration.spec.ts]: Will revert those again, as we loose information of the uri comparison and would only get an boolean assertion output which is way less informative than the real string comparison (see also https://github.com/eclipse-emfcloud/emfcloud-modelserver-theia/pull/121#discussion_r1019568709).

File: packages/modelserver-client/src/model-server-client-api-v1.ts (hunk 7 above)
Comment (ndoschek) [reply packages/modelserver-client/src/model-server-client-api-v1.ts:33]: I would like to stick to one type, that makes things a little more unified. Also the documentation of the URI library speaks of URLs being a "subset" of the URIs according to definition: https://medialize.github.io/URI.js/about-uris.html

File: packages/modelserver-client/src/model-server-client-v2-integration.spec.ts (hunk 9 above)
Comment (eneufeld) [reply packages/modelserver-client/src/model-server-client-v2-integration.spec.ts]: why not use the second parameter for a better failure error?

File: packages/modelserver-client/src/model-server-client-api-v1.ts (hunk 7 above)
Comment (eneufeld) [reply packages/modelserver-client/src/model-server-client-api-v1.ts:33]: Yes URLs are a subset. In the end in doesn't matter. I agree that for code simplification it make sense to stick to one type.

File: packages/modelserver-client/src/model-server-client-v2-integration.spec.ts (hunk 9 above)
Comment (ndoschek) [reply packages/modelserver-client/src/model-server-client-v2-integration.spec.ts]: Oh sorry, I didn't see your comment earlier. As discussed we will stick to the string comparisons for now and will create dedicated URI tests in a follow up (see #123 ) where we should test all varieties (e.g. relative hierarchies, random order of query parameters).

File: examples/dev-example/src/node/model-server-client.ts (hunk 1 above)
Comment (ndoschek) [reply examples/dev-example/src/node/model-server-client.ts:28]: Follow-up ticket: #124
